Documentation Update for 0.3.2 (#71)
CrayLabs/SmartSim: README.md

SmartSim makes it easier to use common Machine Learning (ML) libraries
like PyTorch and TensorFlow, in High Performance Computing (HPC) simulations
and applications.

SmartSim provides an API to connect HPC workloads, particularly (MPI + X) simulations,
to an in-memory database called the Orchestrator, built on an in-memory database
called Redis.

Applications integrated with the SmartRedis clients, written in Fortran, C, C++ and Python,
can stream tensors and datasets to and from the Orchestrator This Client-Server

> paradigm allows for data to be seemlessly exchanged between applications at runtime.

Reviewer: ```suggestion
paradigm allows for data to be seamlessly exchanged between applications at runtime.
```

> In addition to exchanging data between langauges, any of the SmartRedis clients can

Reviewer: ```suggestion
In addition to exchanging data between languages, any of the SmartRedis clients can
```

remotely execute Machine Learning models and TorchScript code on data stored in

> the Orchestrator dispite which langauge the data originated from.

Reviewer: ```suggestion
the Orchestrator despite which language the data originated from.
```

 - Slurm
 - LSF
 - PBSPro
 - Cobalt
 - Local (for laptops/single node, no batch)

### Interactive Launch Example

The following launches the same ``hello_world`` model in an interactive allocation
using the Slurm launcher. Jupyter/IPython notebooks, and scripts

```bash
# get interactive allocation
salloc -N 1 -n 32 --exclusive -t 00:10:00
```

```python
# hello_world.py
from smartsim import Experiment
from smartsim.settings import SrunSettings

exp = Experiment("hello_world_exp", launcher="slurm")
srun = SrunSettings(exe="echo", exe_args="Hello World!")
srun.set_nodes(1)
srun.set_tasks(32)

model = exp.create_model("hello_world", srun)
exp.start(model, block=True, summary=True)

print(exp.get_status(model))
```
```bash
# in interactive terminal
python hello_world.py
```

This script could also be launched in a batch file instead of an
interactive terminal.

```bash
#!/bin/bash
#SBATCH --exclusive
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=32
#SBATCH --time=00:10:00

python /path/to/script.py
```
```bash
# on Slurm system
sbatch run_hello_world.sh
```

### Batch Launch Examples

SmartSim can also launch workloads in a batch directly from Python, without the need

> for a batch script. Users can launch groups of ``Model`` instances in a ``Ensemble``.

Reviewer: ```suggestion
for a batch script. Users can launch groups of ``Model`` instances in an ``Ensemble``.
```

> The following launches 4 replicas of the the same ``hello_world`` model.

Reviewer: ```suggestion
The following launches 4 replicas of the same ``hello_world`` model.
```

from smartsim import Experiment
from smartsim.settings import AprunSettings, QsubBatchSettings

exp = Experiment("hello_world_batch", launcher="pbs")

# define resources for all ensemble members
qsub = QsubBatchSettings(nodes=4, time="00:10:00",
                        account="12345-Cray", queue="cl40")

# define how each member should run
aprun = AprunSettings(exe="echo", exe_args="Hello World!")
aprun.set_tasks(32)

ensemble = exp.create_ensemble("hello_world", batch_settings=qsub,
                                run_settings=aprun, replicas=4)
exp.start(ensemble, block=True, summary=True)

print(exp.get_status(ensemble))
```
```bash
# on PBS system
python hello_ensemble_pbs.py
```

--------

# Infrastructure Library Applications
 - Orchestrator - In-memory data store and Machine Learning Inference (Redis + RedisAI)
 - Ray - Distributed Reinforcement Learning (RL), Hyperparameter Optimization (HPO)

## Redis + RedisAI

The ``Orchestrator`` is an in-memory database that utilizes Redis and RedisAI to provide
a distributed database and access to ML runtimes from Fortran, C, C++ and Python.

SmartSim provides classes that make it simple to launch the database in many

> configurations and optional form a distributed database cluster. The examples

Reviewer: ```suggestion
configurations and optionally form a distributed database cluster. The examples
```

below will show how to launch the database. Later in this document we will show
how to use the database to perform ML inference and processing.

### Local Launch

The following script launches a single database using the local launcher.

```python
# run_db_local.py
from smartsim import Experiment
from smartsim.database import Orchestrator

exp = Experiment("local-db", launcher="local")
db = Orchestrator(port=6780)

# by default, SmartSim never blocks execution after the database is launched.
exp.start(db)

> # launch models, anaylsis, training, inference sessions, etc

Reviewer: ```suggestion
# launch models, analysis, training, inference sessions, etc
```

# that communicate with the database using the SmartRedis clients

# stop the database
exp.stop(db)
```

### Interactive Launch

The ``Orchestrator``, like ``Ensemble`` instances, can be launched locally, in interactive
allocations, or in a batch.

The Orchestrator is broken into several classes to ease submission on
HPC systems.

The following example launches a distributed (3 node) database cluster on
a Slurm system from an interactive allocation terminal.

```bash
# get interactive allocation
salloc -N 3 --ntasks-per-node=1 --exclusive -t 00:10:00
```
```python
# run_db_slurm.py
from smartsim import Experiment
from smartsim.database import SlurmOrchestrator

exp = Experiment("db-on-slurm", launcher="slurm")
db_cluster = SlurmOrchestrator(db_nodes=3, db_port=6780, batch=False)

exp.start(db_cluster)

print(f"Orchestrator launched on nodes: {db_cluster.hosts}")

> # launch models, anaylsis, training, inference sessions, etc

Reviewer: ```suggestion
# launch models, analysis, training, inference sessions, etc
```

# that communicate with the database using the SmartRedis clients

exp.stop(db_cluster)
```
```bash
# in interactive terminal
python run_db_slurm.py
```

Here is the same example on a PBS system

```bash
# get interactive allocation
qsub -l select=3:ppn=1 -l walltime=00:10:00 -q cl40 -I
```
```python
# run_db_pbs.py
from smartsim import Experiment
from smartsim.database import PBSOrchestrator

exp = Experiment("db-on-slurm", launcher="slurm")
db_cluster = PBSOrchestrator(db_nodes=3, db_port=6780, batch=False)

exp.start(db_cluster)

print(f"Orchestrator launched on nodes: {db_cluster.hosts}")

> # launch models, anaylsis, training, inference sessions, etc

Reviewer: ```suggestion
# launch models, analysis, training, inference sessions, etc
```

# that communicate with the database using the SmartRedis clients

exp.stop(db_cluster)
```
```bash
# in interactive terminal
python run_db_pbs.py
```

### Batch Launch

The ``Orchestrator`` can also be launched in a batch without the need for an interactive allocation.
SmartSim will create the batch file, submit it to the batch system, and then wait for the database
to be launched. Users can hit CTRL-C to cancel the launch if needed.

```Python
# run_db_pbs_batch.py
from smartsim import Experiment
from smartsim.database import PBSOrchestrator

exp = Experiment("db-on-slurm", launcher="pbs")
db_cluster = PBSOrchestrator(db_nodes=3, db_port=6780, batch=True,
                             time="00:10:00", account="12345-Cray", queue="cl40")

exp.start(db_cluster)

print(f"Orchestrator launched on nodes: {db_cluster.hosts}")

> # launch models, anaylsis, training, inference sessions, etc

Reviewer: ```suggestion
# launch models, analysis, training, inference sessions, etc
```

SmartRedis clients are available in Fortran, C, C++, and Python.
Users can seamlessly pull and push data from the Orchestrator from different languages.

## Tensors

Tensors are the fundamental data structure for the SmartRedis clients. The Clients
use the native array format of the language. For example, in Python, a tensor is

> a NumPy array. The C++/C client accepts nested and contingous arrays.

Reviewer: ```suggestion
a NumPy array. The C++/C client accepts nested and contiguous arrays.
```

When stored in the database, all tensors are stored in the same format. Hence,

> any language can recieve a tensor from the database no matter what supported language

Reviewer: ```suggestion
any language can receive a tensor from the database no matter what supported language
```

the array was sent from. This enables applications in different languages to communicate
numerical data with each other at runtime (coupling).

For more information on the tensor data structure, see
[the documentation](https://www.craylabs.org/docs/sr_data_structures.html#tensor)

## Datasets

Datasets are collections of Tensors and associated metadata. The ``Dataset`` class
is a user space object that can be created, added to, sent to, and retrieved from
the Orchestrator database.

For an example of how to use the ``Dataset`` class, see the [Online Analysis example](#online-analysis)

For more information on the API, see the
[API documentation](https://www.craylabs.org/docs/sr_data_structures.html#dataset)

## Examples

Even though the clients rely on the Orchestrator database to be running, it can be helpful

> to see examples of how the API is used accross different languages even without the

Reviewer: ```suggestion
to see examples of how the API is used across different languages even without the
```

infrastructure code. The following examples provide simple examples of client usage.

> For more imformation on the SmartRedis clients, see the

Reviewer: ```suggestion
For more information on the SmartRedis clients, see the
```

[API documentation](https://www.craylabs.org/docs/api/smartredis_api.html) and
[tutorials](https://www.craylabs.org/docs/tutorials/smartredis.html).

**Please note** these are client examples, they will not run if there is no database to
connect to.

### Python

Training code and Model construction are not shown here, but the example below

> shows how to take a PyTorch model, sent it to the database, and execute it

Reviewer: ```suggestion
shows how to take a PyTorch model, send it to the database, and execute it
```

```

### Fortran

You can also load a model from file and put it in the database before you execute it.
This example shows how this is done in Fortran.

```fortran
program run_mnist_example

  use smartredis_client, only : client_type
  implicit none

  character(len=*), parameter :: model_key = "mnist_model"
  character(len=*), parameter :: model_file = "../../cpp/mnist_data/mnist_cnn.pt"

  type(client_type) :: client
  call client%initialize(.false.)

  ! Load pre-trained model into the Orchestrator database
  call client%set_model_from_file(model_key, model_file, "TORCH", "GPU")
  call run_mnist(client, model_key)

contains

subroutine run_mnist( client, model_name )
  type(client_type), intent(in) :: client
  character(len=*),  intent(in) :: model_name

  integer, parameter :: mnist_dim1 = 28
  integer, parameter :: mnist_dim2 = 28
  integer, parameter :: result_dim1 = 10

  real, dimension(1,1,mnist_dim1,mnist_dim2) :: array
  real, dimension(1,result_dim1) :: result

  character(len=255) :: in_key
  character(len=255) :: out_key

  character(len=255), dimension(1) :: inputs
  character(len=255), dimension(1) :: outputs

>   ! Construct the keys used for the specifiying inputs and outputs

Reviewer: ```suggestion
  ! Construct the keys used for the specifying inputs and outputs
```

  in_key = "mnist_input"
  out_key = "mnist_output"

  ! Generate some fake data for inference
  call random_number(array)
  call client%put_tensor(in_key, array, shape(array))

  inputs(1) = in_key
  outputs(1) = out_key
  call client%run_model(model_name, inputs, outputs)
  result(:,:) = 0.
  call client%unpack_tensor(out_key, result, shape(result))

end subroutine run_mnist

end program run_mnist_example

```

---------
# SmartSim + SmartRedis

SmartSim and SmartRedis were designed to work together. When launched through

> SmartSim, applcations using the SmartRedis clients are directly connected to

Reviewer: ```suggestion
SmartSim, applications using the SmartRedis clients are directly connected to
```

together.

## Online Analysis

Using SmartSim, HPC applications can be monitored in real time by streaming data
from the application to the database. SmartRedis clients can retrieve the
data, process, analyze it, and store the data in the database.

The following is an example of how a user could monitor and analyze a simulation.
The example here uses the Python client, but SmartRedis clients are available in
C++, C, and Fortran as well and implement the same API.

The example will produce the visualization below while the simulation is running.

https://user-images.githubusercontent.com/13009163/127622717-2c9e4cfd-50f4-4d94-88c4-8c05fa2fa616.mp4

#### Lattice Boltzmann Simulation

Using a [Lattice Boltzmann Simulation](https://en.wikipedia.org/wiki/Lattice_Boltzmann_methods),
this example demonstrates how to use the SmartRedis ``Dataset`` API to stream
data to the Orchestrator deployed by SmartSim.

The following code will show the peices of the simulation that are needed to
transmit the data needed to plot timesteps of the simulation.

```Python
# fv_sim.py
from smartredis import Client
import numpy as np

> # initialization code ommitted

Reviewer: ```suggestion
# initialization code omitted
```

    # Put Dataset in db at key "data_{time_step}"
    client.put_dataset(dataset)
```

The driver that launches the database and the simulation (non-blocking), looks
like:

```Python
# driver.py
time_steps, seed = 3000, 42

exp = Experiment("finite_volume_simulation", launcher="local")
db = Orchestrator(port=6780)
settings = RunSettings("python", exe_args=["fv_sim.py",
                                           f"--seed={seed}",
                                           f"--steps={time_steps}"])
model = exp.create_model("fv_simulation", settings)
model.attach_generator_files(to_copy="fv_sim.py")
exp.generate(db, model, overwrite=True)

exp.start(db)
client = Client(address="127.0.0.1:6780", cluster=False)

# start simulation (non-blocking)
exp.start(model, block=False, summary=True)

# poll until simulation starts and then retrieve data
client.poll_key("cylinder", 200, 100)
cylinder = client.get_tensor("cylinder").astype(bool)

for i in range(0, time_steps):
    client.poll_key(f"data_{str(i)}", 10, 1000)
    dataset = client.get_dataset(f"data_{str(i)}")
    ux, uy = dataset.get_tensor("ux"), dataset.get_tensor("uy")

    # analysis/plotting code omitted

exp.stop(db)
```

> More details about online anaylsis with SmartSim and the full code examples can be found in the

Reviewer: ```suggestion
More details about online analysis with SmartSim and the full code examples can be found in the
```

runtime linked into the database.

Any of the functions available in the
[TorchScript builtins](https://pytorch.org/docs/stable/jit_builtin_functions.html#builtin-functions)
can be saved as "script" or "functions" in the database and used directly by
any of the SmartRedis Clients.

### Singular Value Decomposition

For example, the following code sends the built-in
[Singular Value Decomposition](https://pytorch.org/docs/stable/generated/torch.svd.html)
to the database and execute it on a dummy tensor.

```python
import numpy as np
from smartredis import Client

# don't even need to import torch
def calc_svd(input_tensor):
    return input_tensor.svd()

# connect a client to the database
client = Client(address="127.0.0.1:6780", cluster=False)

# get dummy data
tensor = np.random.randint(0, 100, size=(5, 3, 2)).astype(np.float32)

client.put_tensor("input", tensor)
client.set_function("svd", calc_svd)

client.run_script("svd", "calc_svd", "input", ["U", "S", "V"])
# results are not retrieved immediately in case they need
# to be fed to another function/model

U = client.get_tensor("U")
S = client.get_tensor("S")
V = client.get_tensor("V")
print(f"U: {U}, S: {S}, V: {V}")
```

> The processing capabilties make it simple to form computational piplines of

Reviewer: ```suggestion
The processing capabilities make it simple to form computational pipelines of
```

functions, scripts, and models.

See the full [TorchScript Language Reference](https://pytorch.org/docs/stable/jit.html#torchscript-language)
documentation for more information on available methods, functions, and how
to create your own.

## Online Inference

> SmartSim supports the following frameworks for quering Machine Learning models

Reviewer: ```suggestion
SmartSim supports the following frameworks for querying Machine Learning models
```

the SmartRedis examples in [Python](#Python), [C++](#c), and [Fortran](#fortran).

### TensorFlow and Keras

The Orchestrator, in addition to PyTorch, is built with TensorFlow support by default.
Currently TensorFlow 2.4.0 is supported, but the graph of the model must be frozen
before it is placed in the database.

SmartSim include a utility to freeze the graph of a TensorFlow or Keras model in
``smartsim.tf``

The example below shows how to use the utility to freeze an mnist model created in
Keras. For the sake of brevity, we will use the Python client for further examples,
however, the API is the same for the other clients.

```Python
# create a simple Fully connected network in Keras
model = keras.Sequential(
    layers=[
        keras.layers.InputLayer(input_shape=(28, 28), name="input"),
        keras.layers.Flatten(input_shape=(28, 28), name="flatten"),
        keras.layers.Dense(128, activation="relu", name="dense"),
        keras.layers.Dense(10, activation="softmax", name="output"),
    ],
    name="FCN",
)

# Compile model with optimizer
model.compile(optimizer="adam",
             loss="sparse_categorical_crossentropy",
             metrics=["accuracy"])

> # training code ommited

Reviewer: omitted

)

input_data = np.random.rand(1, 28, 28).astype(np.float32)
client.put_tensor("input", input_data)
client.run_model("keras_fcn", "input", "output")

pred = client.get_tensor("output")
print(pred)
```
------
### ONNX

ONNX is a standard format for representing models. A number of different Machine Learning
Libraries are supported by ONNX and can be readily used with SmartSim.

Some popular ones are:
 - [Scikit-learn](https://scikit-learn.org)
 - [XGBoost](https://xgboost.readthedocs.io)
 - [CatBoost](https://catboost.ai)
 - [TensorFlow](https://www.tensorflow.org)
 - [Keras](https://keras.io)
 - [PyTorch](https://pytorch.org)
 - [LightGBM](https://lightgbm.readthedocs.io/en/latest/)
 - [libsvm](https://www.csie.ntu.edu.tw/~cjlin/libsvm/)

As well as some that are not listed. There are also many tools to help convert
models to ONNX.

 - [onnxmltools](https://github.com/onnx/onnxmltools)
 - [skl2onnx](https://github.com/onnx/sklearn-onnx/)
 - [tensorflow-onnx](https://github.com/onnx/tensorflow-onnx/)

And PyTorch has it's own converter.

Below are some examples of a few models in Sci-kit Learn that are converted

> into onnx format for use with SmartSim.

Reviewer: for consistency, `onnx` -> `ONNX`

#### KMeans

K-means clustering is an unsupervised ML algorithm. It is used to categorize data points

> into f groups ("clusters"). Sci-kit Learn has a built in implementation of K-means clustering

Reviewer: `built in` -> `built-in`

and it is easily converted to ONNX for use with SmartSim through ``skl2onnx.to_onnx()``.

```python

X = np.arange(20, dtype=np.float32).reshape(10, 2)
tr = KMeans(n_clusters=2)
tr.fit(X)

kmeans = to_onnx(tr, X, target_opset=11)
model = kmeans.SerializeToString()

sample = np.arange(20, dtype=np.float32).reshape(10, 2) # dummy data
client.put_tensor("input", sample)

client.set_model("kmeans", model, "ONNX", device="CPU")
client.run_model("kmeans", inputs="input", outputs=["labels", "transform"])

print(client.get_tensor("labels"))
```

#### Random Forest

> The Random Forest example uses the Iris datset from Sci-kit Learn to train a

Reviewer: dataset
